lib/bundler/definition.rb

```diff
@@ -54,7 +54,8 @@ def self.build(gemfile, lockfile, unlock)
     def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, optional_groups = [])
       @unlocking = unlock == true || !unlock.empty?
 
-      @dependencies    = dependencies
+      @dependencies = dependencies
+      @dependencies_by_name = dependencies.group_by(&:name)
```

Comment:
Agree it's not ideal, but I can't think of how else to fix the `only_update_to_newer_versions` bug